api/controllers/clients.js

```diff
@@ -0,0 +1,224 @@
+        throw new Error(`${action} :: Authorization Failed [Invalid Email Or Password]!`);
+    }
+    let client = await getClientByEmail(email);
+    if (!client) 
+        throwAuthError ();
+    if (!(await bcrypt.compare(password, client[PASSWORD])))
+        throwAuthError ();
+    return client;
+}
+
+
+/*
+INPUT: { NONE }
+OUTPUT
+{
+    clients: [ 
+        { < client object with info > },
+        { < client object with info > },
+        ...
+    ]
+}
+*/
+module.exports.get_clients_list = async (req, response, next) => {
+    try {
+        let clients = (await db.query(`SELECT ${CLIENT_ID}, ${DISPLAY_NAME}, ${BIO} FROM ${TABLE} WHERE ${CLIENT_ID} != '${SAFETRACE_CLIENT_ID}';`)).rows;
+        response.status(200).json( { clients: clients } );
+    }
+    catch (e) { next(e); }
+};
+
+
+/*
+INPUT:
+{
+    HEADER
+    {
+        email:      < email >,
+        password:   < password >,
+    }
+    display_name:   < display name >
+    bio:            < a short bio >
+}
+OUTPUT
+{
+    message:        `Account Created For: < email >, save the API and Private Keys included in this object. Make sure to keep them private and secure`,
+    api_key:        < api key >,
+    private_key:    < private key >
+    public_key:     < public key >
+}
+*/
+module.exports.client_signup = async (req, response, next) => {
+    try {
+        const action = 'SIGN UP';
+        let { email, password } = checkForEmailAndPassword (req, action);
+        
+        let bio = assertBodyKey (req, BIO, action);
+        let displayName = assertBodyKey (req, DISPLAY_NAME, action);
+        
+        function throwError () {
+            throw new Error(`${action} :: Invalid Or Existing Email Address: ${email}`);
+        }
+        
+        if (!validEmailRegex.test(email))
+            throwError();
+        
+        if ((await getClientByEmail(email))) 
+            throwError();
+            
+        const passwordHash = await hashing.hashStringRandom(password);
+            
+        // make uuid for api key
+        let apiKey = await v1(); 
+        let encKeys = a_Encryption.getKeyPair();
+
+        let encryptedAPIKey = s_Encryption.encryptData(apiKey, process.env.SAFETRACE_KEY, process.env.SAFETRACE_IV);
+        let encryptedPrivateKey = s_Encryption.encryptData(encKeys.privateKey, process.env.SAFETRACE_KEY, process.env.SAFETRACE_IV);
+
+        await db.query(`INSERT INTO ${TABLE} (${EMAIL}, ${DISPLAY_NAME}, ${PASSWORD}, ${BIO}, ${API_KEY}, ${PUBLIC_KEY}, ${PRIVATE_KEY}) VALUES ('${email}', '${displayName}', '${passwordHash}', '${bio}', '${encryptedAPIKey}', '${encKeys.publicKey}', '${encryptedPrivateKey}');`);
+        
+        // TODO: add a sign in explanation prompt...
+        response.status(201).json( { 
+            message: `Account Created For: ${email}, save the API and Private Keys included in this object. Make sure to keep them private and secure`,
+            api_key: apiKey,
+            private_key: encKeys.privateKey,
```

Comment:
fine for now, but just FYI the "proper" way to make this 100% safe for the end user, is THEY generate the keys on the frontend, and  THEY post the public key only to the backend on signup. That way, the backend isn't in charge of creating the public key, and never saw it (thus you don't have to "trust" the backend to dispose of it and not decrypt all your data)

Reply:
When I say keys here I mean the assymetric keypair.

Reply:
good call, Ill add another 'temporary endpoint' to act as our "encryption library"

Reply:
where we can obtain keys as well

Reply:
sure